File: dungeons-1.21.11/src/main/java/fishmod/features/dungeon/StarredMobHighlight.java
package fishmod.features.dungeon;

import fishmod.utils.Location;
import fishmod.utils.config.values.FishSettings;
import fishmod.utils.rendering.RenderUtils;
import fishmod.utils.rendering.RenderingEvents;
import net.fabricmc.fabric.api.client.rendering.v1.world.WorldRenderContext;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.render.VertexConsumer;
import net.minecraft.client.util.math.MatrixStack;
import net.minecraft.client.world.ClientWorld;
import net.minecraft.entity.Entity;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.decoration.ArmorStandEntity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.util.math.Box;

import java.util.HashSet;
import java.util.Set;

/**
 * Outlines dungeon mobs marked as "starred" elites. Hypixel doesn't put the ✯ on the mob's own
 * nametag — it's on a separate invisible armor stand riding/hovering above the mob that also
 * shows its health (e.g. "Zombie Knight ✯300,000/300,000❤"). So detection works by scanning
 * armor-stand nametags for the star + heart markers, then picking the nearest non-armor-stand
 * living entity underneath as the actual mob to outline (same approach NoammAddons uses).
 */
public final class StarredMobHighlight {

    private static final String STAR = "✯";
    private static final String HEART = "❤";

    private StarredMobHighlight() {}

    public static void init() {
        RenderingEvents.OUTLINE_ENTITY.register(StarredMobHighlight::render);
    }

    private static boolean active() {
        return FishSettings.enableStarredMobHighlight && Location.inDungeon();
    }

    private static void render(WorldRenderContext context, MatrixStack matrices, VertexConsumer consumer) {
        if (!active()) return;
        MinecraftClient mc = MinecraftClient.getInstance();
        ClientWorld world = mc.world;
        if (world == null) return;

        float[] rgba = RenderUtils.toFloats(FishSettings.starredMobHighlightColor);

        for (Entity mob : findStarredMobs(world)) {
            if (mob instanceof LivingEntity living && living.isAlive()) {
                RenderUtils.renderOutline(matrices, consumer, living.getBoundingBox(), rgba);
            }
        }
    }

    private static Set<Entity> findStarredMobs(ClientWorld world) {
        Set<Entity> mobs = new HashSet<>();
        for (Entity entity : world.getEntities()) {
            if (!(entity instanceof ArmorStandEntity stand) || !stand.hasCustomName()) continue;
            String name = stand.getCustomName().getString();
            if (!name.contains(STAR) || !name.contains(HEART)) continue;

            Entity mob = findNearestMob(world, stand);
            if (mob != null) mobs.add(mob);
        }
        return mobs;
    }

    private static Entity findNearestMob(ClientWorld world, ArmorStandEntity stand) {
        Box searchBox = stand.getBoundingBox().expand(1.0, 3.0, 1.0);
        Entity nearest = null;
        double nearestDist = Double.MAX_VALUE;

        for (Entity candidate : world.getOtherEntities(stand, searchBox)) {
            if (!isEligibleMob(candidate)) continue;
            double dist = candidate.squaredDistanceTo(stand);
            if (dist < nearestDist) {
                nearestDist = dist;
                nearest = candidate;
            }
        }
        return nearest;
    }

    private static boolean isEligibleMob(Entity entity) {
        if (entity instanceof ArmorStandEntity) return false;
        if (!(entity instanceof LivingEntity living)) return false;
        if (!living.isAlive()) return false;
        if (entity instanceof PlayerEntity player) return !player.isInvisible();
        return true;
    }
}

File: dungeons-1.21.11/src/main/kotlin/fishmod/features/dungeon/StarredMobHighlight.kt
package fishmod.features.dungeon

import fishmod.utils.Location
import fishmod.utils.config.values.FishSettings
import fishmod.utils.rendering.RenderUtils
import fishmod.utils.rendering.RenderingEvents
import net.fabricmc.fabric.api.client.rendering.v1.world.WorldRenderContext
import net.minecraft.client.MinecraftClient
import net.minecraft.client.render.VertexConsumer
import net.minecraft.client.util.math.MatrixStack
import net.minecraft.client.world.ClientWorld
import net.minecraft.entity.Entity
import net.minecraft.entity.LivingEntity
import net.minecraft.entity.decoration.ArmorStandEntity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.util.math.Box

/** Outlines dungeon mobs marked as "starred" elites. */
object StarredMobHighlight {

    private const val STAR = "✯"
    private const val HEART = "❤"

    @JvmStatic
    fun init() {
        RenderingEvents.OUTLINE_ENTITY.register { context, matrices, consumer -> render(context, matrices, consumer) }
    }

    private fun active(): Boolean {
        return FishSettings.enableStarredMobHighlight && Location.inDungeon()
    }

    private fun render(context: WorldRenderContext, matrices: MatrixStack, consumer: VertexConsumer) {
        if (!active()) return
        val mc = MinecraftClient.getInstance()
        val world = mc.world ?: return

        val rgba = RenderUtils.toFloats(FishSettings.starredMobHighlightColor)

        for (mob in findStarredMobs(world)) {
            if (mob is LivingEntity && mob.isAlive) {
                RenderUtils.renderOutline(matrices, consumer, mob.boundingBox, rgba)
            }
        }
    }

    private fun findStarredMobs(world: ClientWorld): Set<Entity> {
        val mobs: MutableSet<Entity> = HashSet()
        for (entity in world.entities) {
            if (entity !is ArmorStandEntity || !entity.hasCustomName()) continue
            val name = entity.customName!!.string
            if (!name.contains(STAR) || !name.contains(HEART)) continue

            val mob = findNearestMob(world, entity)
            if (mob != null) mobs.add(mob)
        }
        return mobs
    }

    private fun findNearestMob(world: ClientWorld, stand: ArmorStandEntity): Entity? {
        val searchBox = stand.boundingBox.expand(1.0, 3.0, 1.0)
        var nearest: Entity? = null
        var nearestDist = Double.MAX_VALUE

        for (candidate in world.getOtherEntities(stand, searchBox)) {
            if (!isEligibleMob(candidate)) continue
            val dist = candidate.squaredDistanceTo(stand)
            if (dist < nearestDist) {
                nearestDist = dist
                nearest = candidate
            }
        }
        return nearest
    }

    private fun isEligibleMob(entity: Entity): Boolean {
        if (entity is ArmorStandEntity) return false
        if (entity !is LivingEntity) return false
        if (!entity.isAlive) return false
        if (entity is PlayerEntity) return !entity.isInvisible
        return true
    }
}
